### PR TITLE
Check missing headers

### LINKED ISSUE #595: Create CI check to ensure against headers not being installed
This has popped up a couple times and would be nice to be guarded against with a CI check.

### PR DESCRIPTION
**Pull request by chapman39**
This PR adds a script that compares an existing install directory's include dir vs src/serac. The script will be called in a CI job if `CHECK_MISSING_HEADERS` is `"yes"`.

Fixes #595

### CODE REVIEW

**Comment by chapman39** [comment]
When doing some testing I found a header that may be missing:

`Header 'check_gradient.hpp' not found. See CMakeLists.txt in /usr/WS2/meemee/serac/repo/src/serac/numerics/functional/tests`

File: scripts/gitlab/build_and_test.sh [hunk 1]
```diff
@@ -31,3 +31,15 @@ if [[ "$DO_INTEGRATION_TESTS" == "yes" ]] ; then
         exit 1
     fi
 fi
+
+# Check for missing headers
+if [[ "$CHECK_MISSING_HEADERS" == "yes" ]] ; then
+    INSTALL_DIR=*build_and_test_*/install-*
+    SRC_DIR=src
+    python3 scripts/gitlab/check_missing_headers.py -i $INSTALL_DIR -s $SRC_DIR
+    echo "check script returned: >$?<"
```
Comment (chapman39) [inline scripts/gitlab/build_and_test.sh]: this echo will be removed shortly

**Review by samuelpmishLLNL — APPROVED** [review]

**Comment by chapman39** [comment]
its failing but that's expected due to this header that isn't in the install https://lc.llnl.gov/gitlab/smith/serac/-/jobs/1372592#L6888

**Comment by jamiebramwell** [comment]
@white238 , can you take a look at this?

File: scripts/gitlab/check_missing_headers.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#!/bin/sh
+"exec" "python3" "-u" "-B" "$0" "$@"
+##############################################################################
+# Copyright (c) 2023, Lawrence Livermore National Security, LLC and
```
Comment (white238) [inline scripts/gitlab/check_missing_headers.py]: We do a search and replace to update so it needs to match.
```suggestion
# Copyright (c) 2019-2023, Lawrence Livermore National Security, LLC and
```

File: scripts/gitlab/check_missing_headers.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    install_dir = os.path.abspath(install_dir)
+
+    src_dir = args.src_dir
+    if not os.path.exists(src_dir):
+        print("Error: src directory does not exist {}".format(src_dir))
+        return 1
+    src_dir = os.path.abspath(src_dir)
+
+    print("============================================================")
+    print("check_missing_headers.py args")
```
Comment (white238) [inline scripts/gitlab/check_missing_headers.py]: Can we rename this to `check_for_missing_headers.py`

File: .gitlab/build_quartz.yml [hunk 4]
```diff
@@ -41,6 +41,7 @@ quartz-clang_14_0_6-src:
     HOST_CONFIG: "quartz-toss_4_x86_64_ib-${COMPILER}.cmake"
     # TODO: Turn this  back on when input files are working for the Functional-based physics modules
     #DO_INTEGRATION_TESTS: "yes"
+    CHECK_MISSING_HEADERS: "yes"
```
Comment (white238) [inline .gitlab/build_quartz.yml]: This should go in an azure job with the rest of the code quality checks.

**Comment by white238** [comment]
Please fix whatever missing headers in this PR.

**Comment by chapman39** [comment]
@jamiebramwell @white238 ready for another review

**Review by jamiebramwell — APPROVED** [review]

File: scripts/check_for_missing_headers.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+"exec" "python3" "-u" "-B" "$0" "$@"
+##############################################################################
+# Copyright (c) 2019-2023, Lawrence Livermore National Security, LLC and
+# other Serac Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (BSD-3-Clause)
+##############################################################################
+
+"""
+ file: check_missing_headers.py
```
Comment (white238) [inline scripts/check_for_missing_headers.py]:
```suggestion
 file: check_for_missing_headers.py
```

Comment (chapman39) [reply scripts/check_for_missing_headers.py]: thanks for catching this

File: scripts/check_for_missing_headers.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    args, unknown_args = parse_arguments()
+
+    # ensure args are valid
+    install_dir = args.install_dir
+    if not os.path.exists(install_dir):
+        print("Error: install directory does not exist {}".format(install_dir))
+        return 1
+    install_dir = os.path.abspath(install_dir)
+
+    src_dir = args.src_dir
```
Comment (white238) [inline scripts/check_for_missing_headers.py:60]: Please check if `install_dir` and `src_dir` are directories.

File: scripts/check_for_missing_headers.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    # grab headers from install and src
+    install_headers = get_headers_from(os.path.join(install_dir, "include", "serac"))
+    src_headers = get_headers_from(os.path.join(src_dir, "serac"))
+    
+    # check if each header in src is in install as well
+    res = 0
+    for sh in src_headers:
+        found = False
+        for ih in install_headers:
+            if sh["headerfile"] == ih["headerfile"]:
```
Comment (white238) [inline scripts/check_for_missing_headers.py]: What if we have two header files that have the same file name but different paths?

```
/path/one/foo.hpp
/path/two/foo.hpp
```

Comment (chapman39) [reply scripts/check_for_missing_headers.py]: right that would be a problem. i need to compare the relative path from `serac` between the install and src headers.

Comment (white238) [reply scripts/check_for_missing_headers.py]: The relative paths between install and src should always be the same.

Comment (chapman39) [reply scripts/check_for_missing_headers.py]: Tested by copying `mesh_utils.hpp` in another directory:

`Header 'mesh_utils.hpp' missing in /usr/WS2/meemee/serac/repo/src/infrastructure/CMakeLists.txt`

File: scripts/check_for_missing_headers.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        found = False
+        for ih in install_headers:
+            src_relative_header = "{0}/{1}".format(sh["path"], sh["headerfile"])
+            install_relative_header = "{0}/{1}".format(ih["path"], ih["headerfile"])
+            if src_relative_header == install_relative_header:
+                found = True
+                break
+        if not found:
+            cmakelists_path = os.path.join(src_dir, sh["path"], "CMakeLists.txt")
+            print("Header '{0}' missing in {1}".format(sh["headerfile"], cmakelists_path))
```
Comment (white238) [inline scripts/check_for_missing_headers.py]:
```suggestion
            print("Header '{0}' is missing; it should probably be listed in {1}".format(sh["headerfile"], cmakelists_path))
```

Comment (white238) [reply scripts/check_for_missing_headers.py]: the header may be in a nested directory...

**Review by white238 — APPROVED** [review]

File: scripts/check_for_missing_headers.py (hunk 8 above)
Comment (chapman39) [reply scripts/check_for_missing_headers.py]: i see what you mean, thank you

**Pull request merged**